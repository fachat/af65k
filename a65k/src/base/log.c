/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <stdarg.h>
#include <ctype.h>

#include "infiles.h"
#include "log.h"


static err_level level = LEV_WARN;

void log_module_init(err_level l) {
	level = l;
}

void log_set_level(err_level l) {
	level = l;
}


void log_term(const char *msg) {
	
	int newline = 0;

	printf(">>>: ");
	char c;
	while ((c = *msg) != 0) {
		if (newline) {
			printf("\n>>>: ");
		}
		if (isprint(c)) {
			putchar(c);
			newline = 0;
		} else
		if (c == 10) {
			newline = 1;
		} else
		if (c == 13) {
			// silently ignore
		} else {
			printf("{%02x}", ((int)c) & 0xff);
			newline = 0;
		}
		msg++;
	}
	printf("\n");
}

void log_errno(const char *msg, ...) {
       va_list args;
       va_start(args, msg);
	char buffer[1024];

	if (index(msg, '%') != NULL) {
		// the msg contains parameter, so we need to fix it
		vsprintf(buffer, msg, args);
		msg = buffer;
	}
        printf(">> %s: errno=%d: %s\n", msg, errno, strerror(errno));
}

void log_x(err_level lev, const char *msg, ...) {

	const char *prefix = ">>>";
	switch (lev) {
	case LEV_TRACE: prefix = "TRC:"; break;
	case LEV_DEBUG: prefix = "DBG:"; break;
	case LEV_INFO: 	prefix = "INF:"; break;
	case LEV_WARN: 	prefix = "WRN:"; break;
	case LEV_ERROR: prefix = "ERR:"; break;
	case LEV_FATAL: prefix = "FTL:"; break;
	}

	if (lev >= level) {
       		va_list args;
       		va_start(args, msg);

       		printf(prefix);
       		vprintf(msg, args);
		printf("\n");
	}
}




