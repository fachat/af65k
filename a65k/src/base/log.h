/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef LOG_H
#define LOG_H

void log_errno(const char *msg, ...);

void log_error(const char *msg, ...);

#ifdef DEBUG

void log_warn(const char *msg, ...);

void log_info(const char *msg, ...);

void log_debug(const char *msg, ...);

void log_term(const char *msg);

#define	log_rv(rv)	log_error("ERROR RETURN: %d\n", (rv))

#else

#define	log_warn(msg, ...)
#define	log_info(msg, ...)
#define	log_debug(msg, ...)
#define	log_term(msg, ...)
#define	log_rv(rv)

#endif

#endif

