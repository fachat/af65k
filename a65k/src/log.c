/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <stdarg.h>
#include <ctype.h>

void log_term(const char *msg) {
	
	int newline = 0;

	printf(">>>: ");
	char c;
	while ((c = *msg) != 0) {
		if (newline) {
			printf("\n>>>: ");
		}
		if (isprint(c)) {
			putchar(c);
			newline = 0;
		} else
		if (c == 10) {
			newline = 1;
		} else
		if (c == 13) {
			// silently ignore
		} else {
			printf("{%02x}", ((int)c) & 0xff);
			newline = 0;
		}
		msg++;
	}
	printf("\n");
}

void log_errno(const char *msg, ...) {
       va_list args;
       va_start(args, msg);
	char buffer[1024];

	if (index(msg, '%') != NULL) {
		// the msg contains parameter, so we need to fix it
		vsprintf(buffer, msg, args);
		msg = buffer;
	}
        printf(">> %s: errno=%d: %s\n", msg, errno, strerror(errno));
}

void log_warn(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("WRN:");
        vprintf(msg, args);
}

void log_error(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("ERR:");
        vprintf(msg, args);
}

void log_info(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("INF:");
        vprintf(msg, args);
}

void log_debug(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("DBG:");
        vprintf(msg, args);
}



