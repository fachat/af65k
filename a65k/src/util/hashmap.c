/****************************************************************************

    hashmap handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

/*
 * a hashmap is used for example to quickly find labels
 */

#include "mem.h"
#include "array_list.h"
#include "hashmap.h"


static type_t hash_memtype = {
	"hashmap",
	sizeof(hash_t)
};

static type_t hash_bucket_memtype = {
	"hashmap_bucket",
	sizeof(hash_bucket_t)
};

static inline int bucket_from_hash(hash_t *hash, int hashval) {
	// find bucket by computing the modulo of the hash value
	int bucketno = hashval % hash->n_buckets;
	return bucketno;
}

hash_t *hash_init(int approx_size, int nbuckets, 
		int (*hash_from_key)(const void *key), 
		const void* (*key_from_entry)(const void *entry), 
		bool_t (*equals_key)(const void *fromhash, const void *tobeadded)) {

	hash_t *hash = mem_alloc(&hash_memtype);

	hash->approx_size = approx_size;
	hash->n_buckets = nbuckets;
	hash->hash_from_key = hash_from_key;
	hash->key_from_entry = key_from_entry;
	hash->equals_key = equals_key;

	hash->buckets = mem_alloc_n(nbuckets, &hash_bucket_memtype);

	return hash;
}


void *hash_put(hash_t *hash, void *value) {

	void *removed = NULL;

	const void *key = hash->key_from_entry(value);

	// calculate hash
	int hashval = hash->hash_from_key(key);

	// find bucket by computing the modulo of the hash value
	int bucketno = bucket_from_hash(hash, hashval);

	list_t *bucket_list = hash->buckets[bucketno].bucket_list;
	if (bucket_list == NULL) {
		// first with this value
		bucket_list = array_list_init(hash->approx_size / hash->n_buckets);
		hash->buckets[bucketno].bucket_list = bucket_list;
	} else {
		list_iterator_t *iter = list_iterator(bucket_list);
		while (list_iterator_has_next(iter)) {
			void *entry = list_iterator_next(iter);

			if (hash->equals_key(hash->key_from_entry(entry), key)) {
				list_iterator_remove(iter);
				removed = entry;
				break;
			}
		}
		list_iterator_free(iter);
	}
	
	list_add(bucket_list, value);

	return removed;
}


void *hash_get(hash_t *hash, void *key) {

	// calculate hash
	int hashval = hash->hash_from_key(key);

	// find bucket by computing the modulo of the hash value
	int bucketno = bucket_from_hash(hash, hashval);

	list_t *bucket_list = hash->buckets[bucketno].bucket_list;

	if (bucket_list == NULL) {
		return NULL;
	}

	list_iterator_t *iter = list_iterator(bucket_list);

	while (list_iterator_has_next(iter)) {

		void *fromhash = list_iterator_next(iter);
	
		if (hash->equals_key(hash->key_from_entry(fromhash), key)) {

			list_iterator_free(iter);

			return fromhash;
		}
	}

	list_iterator_free(iter);

	return NULL;
}


