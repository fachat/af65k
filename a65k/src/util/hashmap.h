/****************************************************************************

    hashmap handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

/*
 * a hashmap is used for example to quickly find labels
 */

#ifndef HASHMAP_H
#define HASHMAP_H

#include <stdio.h>

#include "types.h"
#include "list.h"


typedef struct hash_s hash_t;

// initialize a hashmap. nbuckets is the number of hash values that are actually used
// The bucket number is computed by taking the modulo of the hash value to the base nbucket.
// The approximate size is divided by nbuckets to determine the initial size of 
// an allocated bucket.
hash_t *hash_init(int approx_size, int nbuckets, 
		int (*hash_from_key)(const void *data), 
		const void* (*key_from_entry)(const void *entry),
		bool_t (*equals_key)(const void *fromhash, const void *tobeadded));

// convenience - key is a string (char *), so we can use our internal functions for 
// hash_from_key and equals_key
// TODO: ignore case flag
hash_t *hash_init_stringkey(int approx_size, int nbuckets, 
		const char* (*key_from_entry)(const void *entry));
// same but ignore case when computing hash and comparing keys
hash_t *hash_init_stringkey_nocase(int approx_size, int nbuckets, 
		const char* (*key_from_entry)(const void *entry));

// adds a new entry; returns any entry that has been removed (if match_equals is set)
void *hash_put(hash_t *, void *value);

// get the value for the given hash key, NULL if none found
void *hash_get(hash_t *hash, const void *key);

static inline bool_t hash_contains(hash_t *hash, void *key) {
        return NULL != hash_get(hash, key);
}



#endif

