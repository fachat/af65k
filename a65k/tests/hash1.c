

#include <string.h>

#define	DEBUG

#include "log.h"
#include "hashmap.h"

typedef struct {
	const char *name;
	const char *data;
} entry_t;

static entry_t data[] = {
	{ "a", "11" }, 
	{ "b", "22" }, 
	{ "c", "33" }, 
	{ "d", "44" }, 
	{ "e", "55" }, 
	{ "f", "66" }, 
	{ "g", "77" }, 
	{ "h", "88" }, 
	{ "i", "99" }
};

static bool_t equals_key(const void *data1, const void *data2) {

	log_debug("equals for '%s' vs. '%s' is %d\n", data1, data2, strcmp((char*)data1, (char*)data2));

	return (0 == strcmp((char*)data1, (char*)data2));
}

static const void* key_from_entry(const void *data) {

	entry_t *entry = (entry_t*) data;

	return 	entry->name;
}


static int hash_from_key(const void *data) {
	
	char *cdata = (char*)data;

	int len = strlen(cdata);
	
	int hash = len;

	// the hash algorithm used below only makes sense
	// for up to 8 bytes ( 13^8 approx. 32 bit )
	if (len > 8) {
		len = 8;
	}
	for (int i = 0; i < len; i++) {
		hash = hash * 13 + cdata[i];
	}

	log_debug("hash for '%s' is %d\n", cdata, hash);

	return hash;
}


void do_test(hash_t *hash);

int main(int argc, char *argv[]) {

	hash_t *hash = NULL;

	log_debug("hashmap:\n");

	hash = hash_init(10, 3, hash_from_key, key_from_entry, equals_key);

	do_test(hash);

	return 0;
}


void do_test(hash_t *hash) {

	for (int i = 0; i < 9; i++) {
		
		hash_put(hash, (void*)&(data[i]));
	}


	entry_t *data = (entry_t*) hash_get(hash, "4");

	log_debug("retrieved '%s'\n",data == NULL ? "<null>" : data->data);

	data = (entry_t*) hash_get(hash, "e");

	log_debug("retrieved '%s'\n",data == NULL ? "<null>" : data->data);

	entry_t new_e = { "e", "66" };

	data = hash_put(hash, &new_e);

	log_debug("removed '%s'\n",data == NULL ? "<null>" : data->data);

	data = (entry_t*) hash_get(hash, "e");

	log_debug("retrieved '%s'\n",data == NULL ? "<null>" : data->data);
}

