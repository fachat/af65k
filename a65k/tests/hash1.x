DBG:hashmap:
DBG:hash for '1' is 62
DBG:hash for '2' is 63
DBG:hash for '3' is 64
DBG:hash for '4' is 65
DBG:hash for '5' is 66
DBG:hash for '6' is 67
DBG:hash for '7' is 68
DBG:hash for '8' is 69
DBG:hash for '9' is 70
DBG:hash for '4' is 65
DBG:equals for '4' vs. '1' is 1
DBG:equals for '4' vs. '4' is 0
DBG:retrieved '4'
