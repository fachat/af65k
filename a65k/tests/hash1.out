DBG:hashmap:
DBG:hash for 'a' is 110
DBG:hash for 'b' is 111
DBG:hash for 'c' is 112
DBG:hash for 'd' is 113
DBG:equals for 'a' vs. 'd' is -1
DBG:hash for 'e' is 114
DBG:equals for 'b' vs. 'e' is -1
DBG:hash for 'f' is 115
DBG:equals for 'c' vs. 'f' is -1
DBG:hash for 'g' is 116
DBG:equals for 'a' vs. 'g' is -1
DBG:equals for 'd' vs. 'g' is -1
DBG:hash for 'h' is 117
DBG:equals for 'b' vs. 'h' is -1
DBG:equals for 'e' vs. 'h' is -1
DBG:hash for 'i' is 118
DBG:equals for 'c' vs. 'i' is -1
DBG:equals for 'f' vs. 'i' is -1
DBG:hash for '4' is 65
DBG:equals for 'a' vs. '4' is 1
DBG:equals for 'd' vs. '4' is 1
DBG:equals for 'g' vs. '4' is 1
DBG:retrieved '<null>'
DBG:hash for 'e' is 114
DBG:equals for 'b' vs. 'e' is -1
DBG:equals for 'e' vs. 'e' is 0
DBG:retrieved '55'
DBG:hash for 'e' is 114
DBG:equals for 'b' vs. 'e' is -1
DBG:equals for 'e' vs. 'e' is 0
DBG:removed '55'
DBG:hash for 'e' is 114
DBG:equals for 'b' vs. 'e' is -1
DBG:equals for 'h' vs. 'e' is 1
DBG:equals for 'h' vs. 'e' is 1
DBG:equals for 'e' vs. 'e' is 0
DBG:retrieved '66'
